pkg/cache/listers.go

```diff
@@ -0,0 +1,154 @@
+package cache
+
+import (
+	"github.com/kcp-dev/apimachinery/pkg/logicalcluster"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+)
+
+// GenericLister is a lister skin on a generic Indexer
+type GenericLister interface {
+	// List will return all objects across clusters
+	List(selector labels.Selector) (ret []runtime.Object, err error)
+	// Get will attempt to retrieve assuming that name==key
+	Get(name string) (runtime.Object, error)
+	// ByCluster will give you a GenericClusterLister for one namespace
+	ByCluster(cluster logicalcluster.LogicalCluster) cache.GenericLister
+}
+
+// NewGenericLister creates a new instance for the genericLister.
+func NewGenericLister(indexer cache.Indexer, resource schema.GroupResource) GenericLister {
+	return &genericLister{indexer: indexer, resource: resource}
+}
+
+type genericLister struct {
+	indexer  cache.Indexer
+	resource schema.GroupResource
+}
+
+func (s *genericLister) List(selector labels.Selector) (ret []runtime.Object, err error) {
+	err = cache.ListAll(s.indexer, selector, func(m interface{}) {
+		ret = append(ret, m.(runtime.Object))
+	})
+	return ret, err
+}
+
+func (s *genericLister) Get(name string) (runtime.Object, error) {
+	obj, exists, err := s.indexer.GetByKey(name)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, errors.NewNotFound(s.resource, name)
+	}
+	return obj.(runtime.Object), nil
+}
+
+func (s *genericLister) ByCluster(cluster logicalcluster.LogicalCluster) cache.GenericLister {
+	return &genericClusterLister{indexer: s.indexer, resource: s.resource, cluster: cluster}
+}
+
+type genericClusterLister struct {
+	indexer  cache.Indexer
+	cluster  logicalcluster.LogicalCluster
+	resource schema.GroupResource
+}
+
+func (s *genericClusterLister) List(selector labels.Selector) (ret []runtime.Object, err error) {
+	selectAll := selector.Empty()
```

Comment:
`nil` is a valid value for labels.Selector, the code I based this off doesn't do anything to handle nil values but I'm assuming we should treat it as equivalent to an empty selector?

Reply:
Sounds good